api/service/log_service.go

```diff
@@ -15,51 +15,239 @@
+	PodName string `json:"pod_name"`
+	// Container name that produces this log
+	ContainerName string `json:"container_name"`
+	// Log in text format
+	TextPayload string `json:"text_payload,omitempty"`
+}
+
+type ReadLogStream struct {
+	Stream        io.ReadCloser
+	PodName       string
+	ContainerName string
+}
+
 type LogService interface {
-	ReadLog(options *LogQuery) (io.ReadCloser, error)
+	StreamLogs(logs chan PodLog, stopCh chan struct{}, options *LogQuery) error
 }
 
 type logService struct {
 	// map of cluster name to cluster client
-	logClients map[string]corev1.CoreV1Interface
+	clusterClients map[string]corev1.CoreV1Interface
 }
 
 // NewLogService create a log service
-// logClients is a map of cluster name to its cluster client (CoreV1Interface)
-func NewLogService(logClients map[string]corev1.CoreV1Interface) LogService {
-	return &logService{logClients: logClients}
+// clusterClients is a map of cluster name to its cluster client (CoreV1Interface)
+func NewLogService(clusterClients map[string]corev1.CoreV1Interface) LogService {
+	return &logService{clusterClients: clusterClients}
 }
 
-// ReadLog read log from a container with the given log options
-// Return ReadCloser object which can be used to stream the log when the LogOptions.Follow is set to true
-func (l logService) ReadLog(options *LogQuery) (io.ReadCloser, error) {
-	logClient, ok := l.logClients[options.Cluster]
+func (l logService) StreamLogs(podLogs chan PodLog, stopCh chan struct{}, options *LogQuery) error {
+	clusterClient, ok := l.clusterClients[options.Cluster]
 	if !ok {
-		return nil, fmt.Errorf("unable to find cluster %s", options.Cluster)
+		return fmt.Errorf("unable to find cluster %s", options.Cluster)
+	}
+
+	namespace := options.Namespace
+
+	labelSelector := l.getLabelSelector(*options)
+	pods, err := clusterClient.Pods(namespace).List(metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return err
 	}
 
-	return logClient.Pods(options.Namespace).GetLogs(options.PodName, options.ToKubernetesLogOption()).Stream()
+	go func() {
+		for {
+			allPodLogs := make([]*PodLog, 0)
+
+			for _, pod := range pods.Items {
+				for _, initContainer := range pod.Spec.InitContainers {
+					options.ContainerName = initContainer.Name
+					stream, err := clusterClient.Pods(namespace).GetLogs(pod.Name, options.ToKubernetesLogOption()).Stream()
+					if err != nil {
+						// Error is handled here by logging it rather than returned because the caller usually does not know how to
+						// handle it. Example of what can trigger ListPodLogs error: while the container is being created/terminated
+						// Kubernetes API server will return error when logs are requested. In such case, it is better to return
+						// empty logs and let the caller retry after the container becomes ready eventually.
+						log.Warnf("Failed to ListPodLogs: %s", err.Error())
+						return
+					}
+
+					scanner := bufio.NewScanner(stream)
+					podLogs := make([]*PodLog, 0)
```

Comment:
any reason using array of pointer?

Reply:
No particular reason. I change it to a normal struct.